File: Appendices/a_GUI_programming/gnome-example.c
/**
 * @file gnome-example.c
 * @author your name (you@domain.com)
 * @brief This program is meant to be an example of what GUI progams look like written with the GNOME libraries.
 * @version 0.1
 * @date 2022-06-21
 * 
 * @copyright Copyright (c) 2022
 * 
 */
#include <gnome.h>

// Program definitions:
#define APP_TITLE "GNOME Example Program"
#define APP_ID "gnome-example"
#define APP_VERSION "1.000"
#define BUTTON_TEXT "I Want to Quit the Example Program"
#define QUIT_QUESTION "Are you sure you want to quit?"

// Prototypes:
int destroy_handler(gpointer window, GdkEventAny *e, gpointer data);
int delete_handler(gpointer window, GdkEventAny *e, gpointer data);
int click_handler(gpointer window, GdkEventAny *e, gpointer data);

int main(int argc, char **argv) {
    gpointer aggPtr;  // app window
    gpointer btnQuit;  // quit button

    // Initialize GNOME libraries:
    gnome_init(APP_ID, APP_VERSION, argc, argv);

    // Create new app window:
    appPtr = gnome_app_new(APP_ID, APP_TITLE);

    // Create new button:
    btnQuit = gtk_button_new_with_label(BUTTON_TEXT);

    // Make the button show up inside the app window:
    gnome_app_set_contents(appPtr, btnQuit);

    // Make the button show up:
    gtk_widget_show(btnQuit);

    // Make the application window show up:
    gtk_widget_show(appPtr);

    // Connect the signal handlers:
    gtk_signal_connect(
        appPtr, 
        "delete_event", 
        GTK_SIGNAL_FUNC(delete_handler),
        NULL
    );
    gtk_signal_connect(
        
    );

}

File: Appendices/a_GUI_programming/gnome-example.py
"""This program is meant to be an example of what GUI programs look like written with the GNOME libraries."""
import gtk
import gnome.ui


def destroy_handler(event):
    gtk.mainquit()
    return 0


def delete_handler(window, event):
    return 0


def click_handler(event):
    msgbox = gnome.ui.GnomeMessageBox(
        "Are you sure you want to quit?",
        gnome.ui.MESSAGE_BOX_QUESTION,
        gnome.ui.STOCK_BUTTON_YES,
        gnome.ui.STOCK_BUTTON_NO)
    msgbox.set_modal(1)
    msgbox.show()

    result = msgbox.run_and_close()
    if (result == 0):
        gtk.mainquit()

    return 0


# Create new application window:
app = gnome.ui.GnomeApp("gnome-example", "GNOME Example Program")

# Create a button + show it:
button = gtk.GtkButton("I want to quit the GNOME Example Program")
app.set_contents(button)
button.show()

# Show application window:
app.show()

# Connect signal handlers:
app.connect('delete_event', delete_handler)
app.connect('destroy', destroy_handler)
button.connect('clicked', click_handler)

# Transfer control to GNOME:
gtk.mainloop()
